---
title: "Estimate immune cell proportions in tumor samples"
author: "Stephanie Hickey"
date: `r Sys.Date()`
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

```{r, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE)
```

```{css, echo=FALSE}
pre, code {white-space:pre !important; overflow-x:auto}
```

# Project Summary

Downloading scRNA-seq data and metadata from primary osteosarcoma in dogs to use as reference data (Ammons et al, Commun Biol. 2024 Apr 24).

Deconvolution using parallel methods MuSiC and SCDC.

# Set up

## Make project directories
```{bash, eval=FALSE}
PROJECT=/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution
mkdir $PROJECT/src
mkdir $PROJECT/data
mkdir $PROJECT/results
```

## Download scRNA-seq reference 
```{bash, eval=FALSE}
mkdir $PROJECT/data/Ammons_scrna
cd $PROJECT/data/Ammons_scrna
wget https://zenodo.org/records/10891255/files/canine_naive_n6_annotated.rds
```

## Get gene lengths

Using [gfttools](http://www.genemine.org/gtftools.php) 
```{bash}
module purge
module load Conda/3

pip install gtftools

cd $PROJECT/data

gtftools -l  gene_lengths_Canis_lupus_familiaris.CanFam3.1.100.txt Canis_lupus_familiaris.CanFam3.1.100.gtf
```

## Make random pseuodbulk samples

**script**

**arguments**

**output**

```{bash}

```

## Normalize bulk and pseudobulk data

**script**

**output**


```{bash}
module purge
module load R-bundle-CRAN/2023.12-foss-2023a

Rscript $PROJECT/src/format_bulk.R
```

```{bash}

```

# Deconvolution analysis

```{bash, eval=FALSE}
module purge
module load R-bundle-CRAN/2023.12-foss-2023a
```

## Deconvolution with DWLS and MuSiC using onmideconv

**script**

**arguments**

echo "omnideconv.R path:" $1
echo "ref.seurat.path:" $2
echo "bulk.mtx.path:" $3
echo "tool:" $4
echo "annotation.col:" $5
echo "batch.col:" $6
echo "results.dir:" $7

**output**

```{bash}
cd $PROJECT/run

for tool in music dwls 
do

  echo $tool
  
  sbatch $PROJECT/src/submit_omnideconv.sh \
    $PROJECT/src/omnideconv.R \
    $PROJECT/data/Ammons_scrna/canine_naive_n6_annotated.rds \
    $PROJECT/data/bulk_tpm_mat.Rds \
    $tool \
    celltype.l1 \
    orig.ident \
    $PROJECT/results
  
done

# pseudobulk
mkdir $PROJECT/results/pseudobulk

for tool in music dwls 
do

  echo $tool
  
  sbatch $PROJECT/src/submit_omnideconv.sh \
    $PROJECT/src/omnideconv.R \
    $PROJECT/data/Ammons_scrna/canine_naive_n6_annotated.rds \
    $PROJECT/data/Ammons_scrna/pseudobulk_tpm_mat.Rds \
    $tool \
    celltype.l1 \
    orig.ident \
    $PROJECT/results/pseudobulk
  
done
  
```

## Deconvolution with SCDC

**script**

**arguments**

echo "SCDC.R path:" $1
echo "ref.seurat.path:" $2
echo "bulk.mtx.path:" $3
echo "tool:" $4
echo "annotation.col:" $5
echo "batch.col:" $6
echo "results.dir:" $7

**output**

```{bash}

sbatch $PROJECT/src/submitSCDC.sh \
  $PROJECT/src/SCDC.R \
  $PROJECT/data/Ammons_scrna/canine_naive_n6_annotated.rds \
  $PROJECT/data/bulk_tpm_mat.Rds \
  $tool \
  celltype.l1 \
  orig.ident \
  $PROJECT/results

# pseudobulk
sbatch $PROJECT/src/submitSCDC.sh \
  $PROJECT/src/SCDC.R \
  $PROJECT/data/Ammons_scrna/canine_naive_n6_annotated.rds \
  $PROJECT/data/Ammons_scrna/pseudobulk_tpm_mat.Rds \
  $tool \
  celltype.l1 \
  orig.ident \
  $PROJECT/results/pseuobulk
    
```

## Deconvolution with scaden

echo "bulk.mtx.path:" $1
echo "ref.data.dir:" $2
echo "ref.data.pattern:" $3
echo "results.dir:" $4

```{bash}
cd $PROJECT/run

# real bulk
sbatch $PROJECT/src/run_scaden.sh \
  $PROJECT/data/bulk_tpm_mat.txt \
  $PROJECT/data/Ammons_scrna/for_scaden \
  "*_cpm.txt" \
  $PROEJCT/results

module purge
module load Conda/3

# pseudobulk with same model
scaden predict \
  --model_dir $PROJECT/data/Ammons_scrna/for_scaden \
  --outname $PROEJCT/results/pseudobulk/scaden_proportions.txt \
  $PROJECT/data/Ammons_scrna/pseudobulk_tpm_mat.txt
```
