---
title: "Deconvolution"
author: "Stephanie Hickey"
date: "2024-10-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Project Summary

Downloading scRNA-seq data and metadata from primary osteosarcoma in dogs to use as reference data (Ammons et al, Commun Biol. 2024 Apr 24).

Deconvolution using parallel methods CIBERTSORTx, MuSiC, and SECRET.

# Set up

## Make project directories
```{bash, eval=FALSE}
PROJECT=/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution
mkdir $PROJECT/src
mkdir $PROJECT/data
mkdir $PROJECT/results
```

## Download scRNA-seq reference 
```{bash, eval=FALSE}
mkdir $PROJECT/data/Ammons_scrna
cd $PROJECT/data/Ammons_scrna
wget https://zenodo.org/records/10891255/files/canine_naive_n6_annotated.rds
```

# Deconvolution analysis

```{bash, eval=FALSE}
module purge
module load R-bundle-CRAN/2023.12-foss-2023a
R --vanilla
```

## Set up

Set paths
```{r}
data.dir = "/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution/data"

results.dir = "/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution/results"
```

Load the bulk RNA-seq data
```{r}
bulk = read.csv(paste0(data.dir, "/HS_MSU_RNAseq Analysis_rawCountsWithAnnotations.csv"))
dim(bulk)
```

Load the single-cell reference
```{r}
library(Seurat)

seuratOb = readRDS(paste0(data.dir,"/Ammons_scrna/canine_naive_n6_annotated.rds"))
seuratOb
```

How many genes are shared between the two data sets?
```{r}
sum(bulk$external_gene_name %in% rownames(seuratOb))
```

## Deconvolution with MuSiC

```{r}
library(MuSiC)
library(tidyverse)
```

### Format input data

MuSiC requires bulk data to be formatted as a matrix.
```{r}
# identify rows with duplicated external_gene_name
dups = bulk$external_gene_name[duplicated(bulk$external_gene_name)]

# how many duplicated row names?  
length(dups)

# remove rows with duplicated external_gene_name
bulk.nodups = 
  bulk %>%
  filter(!external_gene_name %in% dups)

# select external_gene_name as rownames
bulk.expr = bulk.nodups[,c(12:34)]
rownames(bulk.expr) = bulk.nodups$external_gene_name

bulk.mtx = as.matrix(bulk.expr)
```

MuSiC requires scRNA-seq reference data to be formatted as a `SingleCellExperiment`.
```{r}
library(SingleCellExperiment)

# Seurat has a function for converting object type
ref.sce = as.SingleCellExperiment(seuratOb)
```

### Estimate cell type proportions:

The function `music_prop` provides output as a list with elements:

* `Est.prop.weighted`: `data.frame` of MuSiC estimated proportions, subjects by cell types   
* `Est.prop.allgene`: `data.frame` of NNLS estimated proportions, subjects by cell types   
* `Weight.gene`: matrix, MuSiC estimated weight for each gene, genes by subjects   
* `r.squared.full`: vector of R squared from MuSiC estimated proportions for each subject    
* `Var.prop`: matrix of variance of MuSiC estimates   

**results files:**

* `music_results.Rdata`   
* `music_proportions.csv`   
* `music_proportions.png`    

```{r}
music.res =  music_prop(bulk.mtx = bulk.mtx, 
                        sc.sce = ref.sce, 
                        clusters = 'celltype.l1',
                        samples = 'orig.ident')

save(music.res, file = paste0(results.dir, "/music_results.Rdata"))

write.csv(music.res$Est.prop.weighted, file = paste0(results.dir,"/music_proportions.csv"))

# stacked bar
res_df = as.data.frame(music.res$Est.prop.weighted)
res_df$Samples =rownames(res_df)

music_long = 
  res_df %>%
  pivot_longer(cols = -Samples,
               names_to = "CellType", 
               values_to = "Proportion") %>%
  mutate(Tool = "MuSiC",
         CellType = gsub(" ", ".", CellType))

# http://vrl.cs.brown.edu/color
col20 = c("#4f8c9d", "#b9cf84", "#a84d64", "#7ed4d8", "#42455e", "#96ea4e", "#4b3596", "#39eec0", "#f92e5d", "#638123", "#5010c8", "#e4b5ff", "#b34a12", "#f1bb99", "#8d6cf6", "#20f53d", "#ea2db9", "#5281e3", "#31a62e", "#fd8992")

ggplot(music_long, 
       aes(x = Samples, 
           y = Proportion, 
           fill = CellType)) +
  geom_col() +
  scale_fill_manual(values = col20) +
  theme_classic() +
  coord_flip()

ggsave(file = paste0(results.dir, "/music_proportions.png"))
```

## Deconvolution with SCDC

```{r}
library(Biobase)
library(SCDC)
```

### Set up 

`SCDC` takes `ExpressionSet` objects with raw read counts as input. It needs an `ExpressionSet` for each reference individual.

Convert the bulk data to an `ExpressionSet`
```{r}
bulk.eset <- getESET(bulk.mtx,
                     pdata = data.frame(Sample = colnames(bulk.mtx)),
                     fdata = rownames(bulk.mtx))
```

Convert reference  `SeuratObject` into an `ExpressionSet` object.
```{r}

ref.eset <- getESET(seuratOb[["RNA"]]$counts,
                    pdata = seuratOb@meta.data,
                    fdata = rownames(seuratOb))

```

Reference QC
```{r}
all_types = unique(seuratOb@meta.data$celltype.l1)

ref.qc <- SCDC_qc(ref.eset,
                  ct.varname = "celltype.l1",
                  sample = "orig.ident",
                  qcthreshold = 0.7,
                  ct.sub = as.character(all_types))

save(ref.qc, file = paste0(results.dir, "/SCDC_QCdata.Rdata"))
```

### Estimate cell type proportions
```{r}
scdc.res = SCDC_prop(
                  bulk.eset = bulk.eset,
                  sc.eset = ref.qc$sc.eset.qc,
                  ct.varname = "celltype.l1",
                  sample = "orig.ident",
                  ct.sub = as.character(all_types))

save(scdc.res, file = paste0(results.dir, "/SCDC_res.Rdata"))
write.csv(scdc.res$prop.est.mvw, file = paste0(results.dir,"/SCDC_proportions.csv"))

# stacked bar
res_df = as.data.frame(scdc.res$prop.est.mvw)
res_df$Samples =rownames(res_df)

scdc_long = 
  res_df %>%
  pivot_longer(cols = -Samples,
               names_to = "CellType", 
               values_to = "Proportion") %>%
  mutate(Tool = "SCDC",
         CellType = gsub(" ", ".", CellType))

ggplot(scdc_long, 
       aes(x = Samples, 
           y = Proportion, 
           fill = CellType)) +
  geom_col() +
  scale_fill_manual(values = col20) +
  theme_classic() +
  coord_flip()

ggsave(file = paste0(results.dir, "/SCDC_proportions.png"))
```

## Deconvolution with SECRET

### Set up

Secret requires a list of cell-type marker genes. Here, I used the marker genes described in Ammons et al's [Supplementary Data 2: `42003_2024_6182_MOESM5_ESM.csv`]( https://static-content.springer.com/esm/art%3A10.1038%2Fs42003-024-06182-w/MediaObjects/42003_2024_6182_MOESM5_ESM.csv)

```{r}

```


## Summary
```{r}
all_long = rbind(music_long)


library(dplyr)
library(tidyr)
library(purrr)

pairwise_correlations <- function(df, group_var, var_name, value) {
  
  # Ensure the data is in long format and grouped by the grouping variable
  df %>%
    # Spread data into wide format so that each variable is a separate column
    pivot_wider(names_from = {{ var_name }}, values_from = {{ value }}) %>%
    # Group by the group_var (can be any variable like 'group', 'ID', etc.)
    group_by({{ group_var }}) %>%
    # Nest data by group to iterate over each group separately
    nest() %>%
    # Apply pairwise correlation within each group
    mutate(cor_results = map(data, ~{
      data <- .x
      # Get variable names
      var_names <- colnames(data)
      # Get all pairwise combinations of variables
      combs <- combn(var_names, 2, simplify = FALSE)
      
      # Perform pairwise correlations for all combinations
      map_dfr(combs, function(pair) {
        cor_test <- cor.test(data[[pair[1]]], data[[pair[2]]], use = "complete.obs")
        tibble(
          var1 = pair[1],
          var2 = pair[2],
          correlation = cor_test$estimate,
          p_value = cor_test$p.value
        )
      })
    })) %>%
    # Unnest the correlation results to get a flat table
    unnest(cor_results)
}

# Example usage (assuming 'df' is your long data frame):
# result <- pairwise_correlations(df, group_var = group, var_name = variable, value = value)
```


## Session info
```{r}
sessionInfo()
```

