---
title: "Estimate immune cell proportions in tumor samples"
author: "Stephanie Hickey"
date: "`r Sys.Date()`"
output: html_document
---

```{r, echo=FALSE}
knitr::opts_chunk$set(echo = FALSE, message=FALSE, warning = FALSE)
```

# Project Summary

Here we use single-cell RNA-seq data from primary osteosarcoma in 6 treatment-naÃ¯ve dogs ([Ammons et al, Commun Biol. 2024 Apr 24](https://www.nature.com/articles/s42003-024-06182-w)) as a reference data set to estimate immune cell proportions in bulk tumor samples using parallel methods [MuSiC](https://xuranw.github.io/MuSiC/articles/MuSiC.html), [SCDC](https://meichendong.github.io/SCDC/articles/SCDC.html#scdc-pre-process-of-scrna-seq-data-1), and [DWLS](https://cran.r-project.org/web/packages/DWLS/index.html).

The following cell types are included in the reference data. Osteoblast and Osteoclasts were removed.

  1. TAM    
  2. T cell   
  3. Endothelial cell   
  4. T_cycling    
  5. B cell   
  6. Neutrophil    
  7. IFN-TAM   
  8. Fibroblast    
  9. T_IFN   
  10. DC    
  11. NK    
  12. Mast cell   
  13. TIM   

Both MuSiC and SCDC are commonly used tools that take advantage of multi-subject reference data sets by appropriately weighting genes showing cross-subject and cross-cell consistency, though they use slightly different algorithms. DWLS does not consider multiple subjects but shows improved performance compared with MuSiC and SCDC using a single-subject reference in a recent [benchmarking study](https://www.biorxiv.org/content/10.1101/2024.06.10.598226v1.full).

## Comparison of tool performace using pseudobulk samples. 

To test how well these tools work with this reference data set, I created 100 pseudobulk samples by randomly selecting 3K (of ~18K) cells from each of the listed cell types in varying proportions.I've compared the known proportions with the predicted proportions from each tool below.

```{r, fig.width=6, fig.height=4, fig.align = 'center'}
# load libraries
library(tidyverse)
library(patchwork)
library(RColorBrewer)
library(DT)

# set paths 
data.dir = "/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution/data"
results.dir = "/mnt/research/bioinformaticsCore/projects/Yuzbasiyan-Gurkan/BCC119_vyg_deconvolution/results"

dwls_pseudo = readRDS(paste0(results.dir, "/pseudobulk/dwls_deconvolution.rds"))
music_pseudo = readRDS(paste0(results.dir, "/pseudobulk/music_deconvolution.rds"))
#scaden_pseudo = read.delim(paste0(results.dir, "/pseudobulk/scaden_proportions.txt"), row.names = 1)

load(paste0(results.dir, "/pseudobulk/SCDC_res.Rdata"))
scdc_pseudo = scdc.res$prop.est.mvw

ground_truth = readRDS(paste0(data.dir, "/Ammons_scrna/pseudobulk/pseudobulk_real_cell_type_proportions.rds"))
samples = ground_truth$Sample
ground_truth$Sample = NULL
ground_truth = as.matrix(ground_truth)
rownames(ground_truth) = samples

make_long = function(wide_mat, tool){
  
  wide_df = as.data.frame(wide_mat)
  wide_df$Samples = rownames(wide_df)
  
  long_df = 
    wide_df %>%
    pivot_longer(cols = -Samples,
                names_to = "CellType", 
                values_to = "Proportion") %>%
    mutate(Tool = tool,
          CellType = gsub(" ", ".", CellType))
  
  return(long_df)
}

plot_propotions = function(long_df, tool){
  
  # http://vrl.cs.brown.edu/color
  col20 = c("#4f8c9d", "#b9cf84", "#a84d64", "#7ed4d8",
            "#42455e", "#96ea4e", "#4b3596", "#39eec0",
            "#f92e5d", "#638123", "#5010c8", "#e4b5ff",
            "#b34a12", "#f1bb99", "#8d6cf6", "#20f53d", 
            "#ea2db9", "#5281e3", "#31a62e", "#fd8992")

  ggplot(long_df, 
        aes(x = Samples, 
            y = Proportion, 
            fill = CellType)) +
    geom_col() +
    ggtitle(tool) + 
    scale_fill_manual(values = col20) +
    theme_classic() +
    coord_flip()
  
  ggsave(file = paste0(results.dir, "/", tool, "_proportions.png"))
  
}

dwls_long = make_long(dwls_pseudo, "DWLS_predicted")
music_long = make_long(music_pseudo, "MuSiC_predicted")
scdc_long = make_long(scdc_pseudo, "SCDC_predicted")
#scaden_long = make_long(scaden_pseudo, "scaden_predicted")
ground_long = make_long(ground_truth, "ground_truth")

all_pseudo = rbind(dwls_long, 
                   music_long,
                   scdc_long,
                   ground_long)

# plot correlations

truth_corr =   
all_pseudo %>%
  pivot_wider(id_cols = c(Samples, CellType), 
              names_from = Tool, 
              values_from = Proportion) %>%
    # Group by the group_var (can be any variable like 'group', 'ID', etc.)
    group_by(Samples) %>%
    # Nest data by group to iterate over each group separately
    nest() %>%
    # Apply pairwise correlation within each group
    mutate(cor_results = map(data, ~{
      data <- .x
      # Get variable names
      var_names <- colnames(data)
      # Get all pairwise combinations of variables
      tools = unique(all_pseudo$Tool)
      combs <- combn(tools, 2, simplify = FALSE)
      combs_keep = keep(combs, function(x) "ground_truth" %in% x)
      # Perform pairwise correlations for all combinations
      map_dfr(combs_keep, function(pair) {
        cor_test <- cor.test(data[[pair[1]]], data[[pair[2]]], use = "complete.obs")
        tibble(
          var1 = pair[1],
          var2 = pair[2],
          correlation = cor_test$estimate,
          p_value = cor_test$p.value
        )
      })
    })) %>%
    # Unnest the correlation results to get a flat table
    unnest(cor_results)

ggplot(truth_corr, aes(x = var1, y = correlation)) +
  geom_boxplot() +
  xlab("Tool used for prediction") +
  ylab("Correlation with ground truth") +
  labs(caption= "A box plot showing the distribution of the Pearson correlations bewteen \nthe predicted and ground-truth cell-type proportions for 100 pseudobulk samples") +
  theme_classic() +
  theme(plot.caption = element_text(hjust = 0)) 

knitr::kable(truth_corr %>%
               select(var1, correlation) %>%
               ungroup() %>%
               group_by(var1) %>%
               summarize(MedianCorrelation = median(correlation),
                         MeanCorrelation = mean(correlation)))
  
```

Cell-type proportion predictions made using DWLS show the best correlation with the ground truth cell-type proportions for each sample. 

## Sample deconvolution using DWLS
```{r, fig.height=7, fig.align = 'center'}
# load predicted cell type proportions
dwls_prop = readRDS(paste0(results.dir, "/dwls_deconvolution.rds"))
write.csv(dwls_prop, file = paste0(results.dir, "/dwls_deconvolution.csv"))

dwls_long = make_long(dwls_prop, "DWLS")
plot_propotions(dwls_long, "DWLS")

 col20 = c("#4f8c9d", "#b9cf84", "#a84d64", "#7ed4d8",
            "#42455e", "#96ea4e", "#4b3596", "#39eec0",
            "#f92e5d", "#638123", "#5010c8", "#e4b5ff",
            "#b34a12", "#f1bb99", "#8d6cf6", "#20f53d", 
            "#ea2db9", "#5281e3", "#31a62e", "#fd8992")

  ggplot(dwls_long, 
        aes(x = Samples, 
            y = Proportion, 
            fill = CellType)) +
    geom_col() +
    ggtitle("DWLS") + 
    scale_fill_manual(values = col20) +
    theme_classic() +
    coord_flip()

dwls_long$Samples = as.factor(dwls_long$Samples)
dwls_long$CellType = as.factor(dwls_long$CellType)
datatable((dwls_long %>% mutate(Proportion = round(Proportion, digits = 2))), filter = "top")
```

## Session info
```{r}
sessionInfo()
```
